sda/protocol.go

```diff
@@ -100,3 +101,96 @@ func ProtocolNameToID(name string) ProtocolID {
 func GlobalProtocolRegister(name string, protocol NewProtocol) (ProtocolID, error) {
 	return protocols.Register(name, protocol)
 }
+
+// ProtocolIO is an interface that allows one protocol to completely define its
+// wire protocol format while still using the Overlay.
+// Cothority sends different messages dynamically as slices of bytes, whereas
+// Google proposes to use union-types:
+// https://developers.google.com/protocol-buffers/docs/techniques#union
+// This is a wrapper to enable union-types while still keeping compatibility with
+// the dynamic cothority-messages. Implementations must provide methods to
+// pass from the 'union-types' to 'cothority-dynamic-messages' with the Wrap
+// and Unwrap method.
+// A default one is provided with defaultProtocolIO so the regular wire-format
+// protocol can still be used.
+type ProtocolIO interface {
+	// Wrap takes a message and the overlay information and returns the message
+	// that has to be sent directly to the network alongside with any error that
+	// happened.
+	// If msg is nil, it is only an internal message of the Overlay.
+	Wrap(msg interface{}, info *OverlayMessage) (interface{}, error)
+	// Unwrap takes the message coming from the network and returns the
+	// inner message that is going to be dispatched to the ProtocolInstance, the
+	// OverlayMessage needed by the Overlay to function correctly and then any
+	// error that might have occurred.
+	Unwrap(msg interface{}) (interface{}, *OverlayMessage, error)
+	// PacketType returns the packet type ID that this Protocol expects from the
+	// network. This is needed in order for the Overlay to receive those
+	// messages and dispatch them to the correct ProtocolIO.
+	PacketType() network.PacketTypeID
+
+	// Name returns the name associated with this ProtocolIO. When creating a
+	// protocol, if one use a name used by a ProtocolIO, this ProtocolIO will be
+	// used to Wrap and Unwrap messages.
+	Name() string
+}
+
+// NewProtocolIO is a function typedef to instantiate a new ProtocolIO.
+type NewProtocolIO func() ProtocolIO
+
+type protocolIOFactoryStruct struct {
+	factories []NewProtocolIO
+}
+
+var protocolIOFactory = protocolIOFactoryStruct{}
+
+// RegisterProtocolIO saves a new NewProtocolIO under its name.
+// When a Conode is instantiated, all ProtocolIOs will be generated and stored
+// for this Conode.
+func RegisterProtocolIO(n NewProtocolIO) {
+	protocolIOFactory.factories = append(protocolIOFactory.factories, n)
+}
+
+// protocolIOStore contains all created ProtocolIOs. It contains the default
+// ProtocolIO used by the Overlay for backwards-compatibility.
+type protocolIOStore struct {
+	sync.Mutex
+	protos    []ProtocolIO
+	defaultIO ProtocolIO
```

Comment:
Here's the slice solution that we discussed then. But I think keeping the defaultIO explicit rather than using a "trick" of using the "0" position in a slice is still better. 
```
for _,p := range protos {
....
}
return defaultIO
```
vs 
```
for _, p := range protos[1:] {
...
}

return protos[0]
```